rust-runtime/aws-smithy-client/src/timeout.rs

```diff
@@ -132,96 +142,39 @@ impl<S> Layer<S> for TimeoutLayer {
-    /// timeout it was.
-    pub enum TimeoutServiceFuture<F> {
-        /// A wrapper around an inner future that will output an [`SdkError`] if it runs longer than
-        /// the given duration
-        Timeout {
-            #[pin]
-            future: Timeout<F, Sleep>,
-            kind: &'static str,
-            duration: Duration,
-        },
-        /// A thin wrapper around an inner future that will never time out
-        NoTimeout {
-            #[pin]
-            future: F
-        }
-    }
-}
-
-impl<F> TimeoutServiceFuture<F> {
-    /// Given a `future`, an implementor of `AsyncSleep`, a `kind` for this timeout, and a `duration`,
-    /// wrap the `future` inside a [`Timeout`] future and create a new [`TimeoutServiceFuture`] that
-    /// will output an [`SdkError`] if `future` doesn't complete before `duration` has elapsed.
-    pub fn new(future: F, params: &TimeoutServiceParams) -> Self {
-        Self::Timeout {
-            future: Timeout::new(future, params.async_sleep.sleep(params.duration)),
-            kind: params.kind,
-            duration: params.duration,
-        }
-    }
-
-    /// Create a [`TimeoutServiceFuture`] that will never time out.
-    pub fn no_timeout(future: F) -> Self {
-        Self::NoTimeout { future }
-    }
-}
-
-impl<InnerFuture, T, E> Future for TimeoutServiceFuture<InnerFuture>
-where
-    InnerFuture: Future<Output = Result<T, SdkError<E>>>,
-{
-    type Output = Result<T, SdkError<E>>;
-
-    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        let (future, kind, duration) = match self.project() {
-            TimeoutServiceFutureProj::NoTimeout { future } => return future.poll(cx),
-            TimeoutServiceFutureProj::Timeout {
-                future,
-                kind,
-                duration,
-            } => (future, kind, duration),
-        };
-        match future.poll(cx) {
-            Poll::Ready(Ok(response)) => Poll::Ready(response),
-            Poll::Ready(Err(_timeout)) => Poll::Ready(Err(SdkError::timeout_error(
-                RequestTimeoutError::new(kind, *duration),
-            ))),
-            Poll::Pending => Poll::Pending,
-        }
-    }
-}
-
-impl<H, R, InnerService, E> tower::Service<Operation<H, R>> for TimeoutService<InnerService>
+impl<Req, S, E> tower::Service<Req> for TimeoutService<S>
 where
-    InnerService: tower::Service<Operation<H, R>, Error = SdkError<E>>,
+    S: tower::Service<Req, Error = SdkError<E>>,
+    // 'static is required because we need to `Box::pin` this future
+    S::Future: 'static,
 {
-    type Response = InnerService::Response;
-    type Error = aws_smithy_http::result::SdkError<E>;
-    type Future = TimeoutServiceFuture<InnerService::Future>;
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>>>>;
```

Comment:
So the tradeoff is that we remove a bunch of code in exchange for adding an allocation per request?